docs/engineering_design.rst

```diff
@@ -3,8 +3,146 @@
+will be revisited in future release to be either modified or completely
+removed.
+
+ * **Multi-tenancy:** One of the main assumptions is that the platform
+   uses the multi-tenancy (Django Site) features without
+   having a main site of its own.
+ * **Dependency on edx-organizations:** In order to implement a
+   full-featured multi-tenancy app, the ``edx-organizations`` is the go-to
+   app to implement this feature.
+ * **Access Control Backends:**
+   Which allows plugins like the CAG app to
+   modify the behaviour of ``courseware.access.has_access`` function
+   in Open edX platform. More info is one the
+   `Access Control Backends`_ pull request.
+   pull request.
+ * **Other app changes:** In order for the app to work it
+   requires three hard-coded changes into the platform that would eventually
+   be upstreamed. For more information see :ref:``supported_open_edx_version``.
+
+
+
+Course Access Groups Database Modules
+-------------------------------------
+
+Naturally, ``course_access_groups.models`` module has the most up-to date
+information and documentation about the database models.
+
+Nevertheless, this section aims to provide a summary of the models
+design. As of writing this docs, there are five distinct new models in this
+app. However, when looking at the logical design we would find only three
+models like the following:
+
+The ``CourseAccessGroup`` Model
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This is the main model which specify which learners should have
+access to which courses.
+
+This model has two Many to Many relationships:
+
+ * **Course:** Courses that belongs to this group. To simplify the use
+   relation with the Django REST Framewor it has been moved to its
+   own model named ``GroupCourse``.
+
+ * **User:** Users who are member of this group. For the same reasons above
+   it has it's own model named ``Membership``.
+
+This model is the only essential model for this app to run. The others can
+be thought as complementary.
+
+The ``PublicCourse`` Model
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+This model is used to mark courses as public to exempt from the
+Course Access Group rules.
+
+Instead of modifying ``CourseOverview`` to have an additional boolean field
+e.g. ``CourseOverview.cag_is_public``, a new model has been created.
+
+This necessary to avoid having complex database migration design and
+minimize the maintenance impact of this app.
+
+The ``MembershipRule`` Model
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+This model is used to automatically assign learners to Course Access Groups.
+It utilizes the
+`USER_ACCOUNT_ACTIVATED <https://github.com/edx/edx-platform/pull/23296>`_
+Django signal to match learners into groups based on their email domain name.
+
+
+Future Plans
+------------
+
+This application works on a *modified* fork of the Hawthorn release. This
+section denotes plans to support future releases the impact on the
+architecture of this app.
+
+Supporting Bridgkeeper in Juniper
```

Comment:
```suggestion
Supporting Bridgekeeper in Juniper
```

Not familiar with it, but *assuming* that's a typo.

Reply:
It is :)